Initial draft for documenting the repository (#1)
j-rivero/ign_moveit2_examples: docs/README.md

# Example documentation

This document describes the different software components

> used by the MoveIt 2, ROS 2, Ignition Gazebo demo,

Reviewer: Can we still use the word Ignition? Maybe link to https://gazebosim.org/ for clarification? Or upstream, they did `~~Ignition~~ Gazebo`.

where to find them and how they are being called.

> ***Note:*** only the follow target is used in this document

Reviewer: ```suggestion
***Note:*** only the Follow Target example is used in this document
```

The document details:

 1. Code repositories being used in the demo
 1. How to prepare a Robot URDF for using Gazebo with ros2_control
 1. How to connect MoveIt 2 with the Gazebo ros2_control plugins
 1. How this example declares the target pose for the manipulation

>  1. Other important ROS2 packages for the simulation

Reviewer: ```suggestion
 1. Other important ROS 2 packages for the simulation
```

 1. Graph of runtime ROS 2 topics and nodes

## Code repositories used in the demo 

Repositories used:

* [Panda Ign MoveIt 2](https://github.com/AndrejOrsula/panda_ign_moveit2):
  Software packages for Franka Emika Panda Robot that enable manipulation with MoveIt 2 
  inside Ignition Gazebo. For control, ignition_ros2_control is used.

> * [ROS Ignition Gazebo](https://github.com/gazebosim/ros_gz/tree/galactichttps://github.com/gazebosim/ros_gz/tree/galactic)

Reviewer: Typo
```suggestion
* [ROS Ignition Gazebo](https://github.com/gazebosim/ros_gz/tree/galactic)
```

>   This repository holds packages that provide integration between ROS and Ignition.

Reviewer: ```suggestion
  This repository holds packages that provide integration between ROS and Gazebo.
```

  Mainly `ros_ign_package` is being used to launch Gazebo with ROS 2 integration 
  and `ros_ign_bridge` to convert Gz messages to ROS messages.

> * [Igition Gazebo ROS2 Control](https://github.com/ros-controls/gazebo_ros2_control/tree/galactic)

Reviewer: ```suggestion
* [Ignition Gz ROS 2 Control](https://github.com/ros-controls/gz_ros2_control) for the latest [Gazebo](https://gazebosim.org) (not to be confused with [Gazebo ROS 2 Control](https://github.com/ros-controls/gazebo_ros2_control/tree/galactic) for [Gazebo-classic](https://classic.gazebosim.org)!)
```
I think you want gz_ros2_control (as opposed to gazebo_ros2_control for Gz-classic) here, which is renamed from ign_ros2_control.

And here I see why you might want to say Ignition or "latest" Gazebo as that package says... sigh. Maybe best to be explicit to eliminate confusion. Maybe ~~Ignition~~ Gazebo again.

  This is a ROS 2 package for integrating the ros2_control controller architecture
  with the Gazebo simulator.

### How to prepare a Robot URDF for using Gazebo with ros2_control

Two main actions needs to be done to prepare the URDF model to use Gazebo and

> ros2_control, configure ros2_control setting using the `ros2_control` URDF

Reviewer: ```suggestion
ros2_control: configure ros2_control setting using the `ros2_control` URDF
```

> macro and add a plugin to Gazebo to parses the `ros2_control` tags and 

Reviewer: ```suggestion
macro, and add a plugin to Gazebo to parse the `ros2_control` tags and 
```

> loads the appropriate hardware interfaces and controller manager

Reviewer: ```suggestion
load the appropriate hardware interfaces and controller manager.
```

#### 1. URDF declaration for the [`ros2_control` URDF tag](https://control.ros.org/master/doc/getting_started/getting_started.html#hardware-description-in-urdf)

This is done by adding the xacro macro `ros2_control_panda_arm` (which is 

> integrated from main Panda arm URDF file) and uses the

Reviewer: ```suggestion
integrated from main Panda arm URDF file) and using the
```

> [`ign_ros2_control/IgnitionSystem`](https://github.com/ros-controls/gz_ros2_control/blob/master/README.md?plain=1#L93-L118):

Reviewer: ```suggestion
[`gz_ros2_control/IgnitionSystem`](https://github.com/ros-controls/gz_ros2_control/blob/master/README.md?plain=1#L93-L118):
```

* https://github.com/AndrejOrsula/panda_ign_moveit2/blob/master/panda_description/urdf/panda.ros2_control#L28-L30

#### 2. URDF declaration for the Gazebo plugin [IgnitionROS2ControlPlugin](https://github.com/ros-controls/gz_ros2_control/blob/master/README.md?plain=1#L153-L169)

This is done by adding the xacro macro `ign_ros2_control` (which is integrated 

> from main Panda URDF file) and add the Gazebo plugin `IgnitionROS2ControlPlugin`

Reviewer: ```suggestion
from main Panda URDF file) and adding the Gazebo plugin `IgnitionROS2ControlPlugin`
```

that parses the `ros2_control` tags and loads the appropriate hardware interfaces and controller manager

> * https://github.com:drejOrsula/panda_ign_moveit2/blob/master/panda_description/urdf/panda.gazebo#L8-L16

Reviewer: ```suggestion
* https://github.com/AndrejOrsula/panda_ign_moveit2/blob/master/panda_description/urdf/panda.gazebo#L8-L16
```

### How to connect MoveIt 2 with the Gazebo ros2_control plugins

> The example launches MoveIt using igniton_ros2_control:

Reviewer: ```suggestion
The example launches MoveIt using gz_ros2_control:
```

* [default.launch.py](../launch/default.launch.py#L81-L94)

This snippet is calling the main [Panda MoveIt configuration launch file](https://github.com/AndrejOrsula/panda_ign_moveit2/blob/master/panda_moveit_config/launch/move_group.launch.py):

* Panda main MoveIt 2 move_group invocation:
  * https://github.com/AndrejOrsula/panda_ign_moveit2/blob/5679c0a97d5e34199271ed7abf02cf303d0e8fb9/panda_moveit_config/launch/move_group.launch.py#L248-L264

* Controller manager
  * https://github.com/AndrejOrsula/panda_ign_moveit2/blob/5679c0a97d5e34199271ed7abf02cf303d0e8fb9/panda_moveit_config/launch/move_group.launch.py#L309-L324

> ### How this example declares the target pose for the manipulation

Reviewer: ```suggestion
### How the Follow Target example declares the target pose for the manipulation
```

This repository uses a `Gazebo PosePublisher` plugin to indicate the target pose for the manipulation:

> * [follow_target.sdf](../ign_moveit2_examples/worlds/follow_target.sdf#L99-L104)

Reviewer: ```suggestion
* [follow_target.sdf](../worlds/follow_target.sdf#L99-L104)
```
Or do we want to use permanlink like all the previous links? Since line numbers are referenced.

Same for the 2 links below.

> Translate the pose from Gazebo to ROS using the `ros_ign_bridge`:

Reviewer: ```suggestion
Translate the pose from Gazebo to ROS using the `ros_gz_bridge`:
```

I don't know. I see that the code says ros_ign_bridge, but I feel like inconsistency will confuse people. Maybe the code should be changed too. If we decide to maintain this as an official example, we should eventually update all the names in code.

* [world_follow_target.launch.py](../launch/worlds/world_follow_target.launch.py#L61-L76)

> The /target_pose is subscribed to by the ROS C++ node that communicates to MoveIt

Reviewer: ```suggestion
The `/target_pose` topic is subscribed to by the ROS C++ node that communicates with MoveIt
```

* [ex_follow_target.cpp](../examples/cpp/ex_follow_target.cpp#L28-L57)

### Other important ROS 2 packages

#### Robot state publisher

The Robot State Publisher is a node and a class to publish
the state of a robot to tf2. At startup time, Robot State Publisher is 
supplied with a kinematic tree model (URDF) of the robot. It then subscribes 

> to the joint_states topic (of type sensor_msgs/msg/JointState) to get

Reviewer: ```suggestion
to the `joint_states` topic (of type `sensor_msgs/msg/JointState`) to get
```

individual joint states. More information in the

> [ROS2 tutorial](https://docs.ros.org/en/galactic/Tutorials/Intermediate/URDF/Using-URDF-with-Robot-State-Publisher.html)

Reviewer: ```suggestion
[ROS 2 tutorial](https://docs.ros.org/en/galactic/Tutorials/Intermediate/URDF/Using-URDF-with-Robot-State-Publisher.html)
```

* https://github.com/ros/robot_state_publisher

#### TF2 

tf2 is the second generation of the transform library, which lets the user
keep track of multiple coordinate frames over time. tf2 maintains the 
relationship between coordinate frames in a tree structure buffered in 
time, and lets the user transform points, vectors, etc between any 
two coordinate frames at any desired point in time.

* http://wiki.ros.org/tf2

## ROS 2 Topics and Nodes

![ROS 2 Graph of the demo](ros_rqt_graph.png)

> Colors legend for nodes:

Reviewer: ```suggestion
Color legend for nodes:
```

 * Green: Gazebo special nodes for ROS integration

>  * Purple: pose target created by the own example code in this repo

Reviewer: ```suggestion
 * Purple: pose target created by the example code in this repo
```

>  * Blue: nodes create and handle by `ros2_control` invoked from `ignition_ros_control`

Reviewer: ```suggestion
 * Blue: nodes created and handled by `ros2_control` invoked from `gz_ros_control`
```

Reviewer: If we keep it ignition, that's fine, but it should be `ign_ros2_control` for consistency.

# Example documentation

> This document describes the different software components
> used by the MoveIt 2, ROS 2 and Ignition Gazebo demo,
> where to find them and how they are being called.

Reviewer 2: ```suggestion
This document describes the different software components
used by the MoveIt 2, ROS 2 and Gazebo demo,
where to find them and how they are being called.
```

> This example and documentation uses the name of ***"Ignition Gazebo"***

Reviewer 2: I think we can mention here the renaming but we should use Gazebo

Author: https://github.com/j-rivero/ign_moveit2_examples/pull/1#issuecomment-1350082938

to refer to the new generation of [Gazebo](https://gazebosim.org/) not
be confused with [Gazebo classic](https://classic.gazebosim.org) also
known as Gazebo11. The Ignition Gazebo project was renamed to just
Gazebo or Gz so the document contains some references to the new URLs
and name like the ones for

> [`gz_ros2_control`](https://github.com/ros-controls/gz_ros2_controlhttps://github.com/ros-controls/gz_ros2_control)

Reviewer 2: ```suggestion
[`gz_ros2_control`](https://github.com/ros-controls/gz_ros2_control)
```

> that should not be confused with the ros2_control wrapper for the
> previous Gazebo classic named

Reviewer 2: ```suggestion
that should not be confused with the `ros2_control` wrapper for the
previous Gazebo classic named
```

> [`gazebo2_ros_control`](https://github.com/ros-controls/gazebo_ros2_control).

Reviewer: ```suggestion
[`gazebo_ros2_control`](https://github.com/ros-controls/gazebo_ros2_control).
```

***Note:*** only the Follow Target example is used in this document

The document details:

 1. Code repositories being used in the demo

>  1. How to prepare a Robot URDF for using Gazebo with ros2_control

Reviewer 2: ```suggestion
 1. How to prepare a Robot URDF for using `Gazebo` with `ros2_control`
```

>  1. How to connect MoveIt 2 with the Gazebo ros2_control plugins

Reviewer 2: ```suggestion
 1. How to connect `MoveIt 2` with the `Gazebo` `ros2_control` plugins
```

 1. How this example declares the target pose for the manipulation
 1. Other important ROS 2 packages for the simulation
 1. Graph of runtime ROS 2 topics and nodes

## Code repositories used in the demo

Repositories used:

* [Panda Ign MoveIt 2](https://github.com/AndrejOrsula/panda_ign_moveit2):
  Software packages for Franka Emika Panda Robot that enable manipulation with MoveIt 2

>   inside Ignition Gazebo. For control, ignition_ros2_control is used.

Reviewer: For consistency with other places in this document
```suggestion
  inside Ignition Gazebo. For control, ign_ros2_control is used.
```

* [ROS Ignition Gazebo](https://github.com/gazebosim/ros_gz/tree/galactic)
  This repository holds packages that provide integration between ROS and Ignition.
  Mainly `ros_ign_package` is being used to launch Gazebo with ROS 2 integration
  and `ros_ign_bridge` to convert Gz messages to ROS messages.

> * [Ignition Gz ROS 2 Control](https://github.com/ros-controls/gz_ros2_control) for the latest [Gazebo](https://gazebosim.org) (not to be confused with [Gazebo ROS 2 Control](https://github.com/ros-controls/gazebo_ros2_control/tree/galactic) for [Gazebo-classic](https://classic.gazebosim.org)!)

Reviewer 2: ```suggestion
* [Gz ROS 2 Control](https://github.com/ros-controls/gz_ros2_control) for the latest [Gazebo](https://gazebosim.org) (not to be confused with [Gazebo ROS 2 Control](https://github.com/ros-controls/gazebo_ros2_control/tree/galactic) for [Gazebo-classic](https://classic.gazebosim.org)!)
```

>   This is a ROS 2 package for integrating the ros2_control controller architecture
>   with the Gazebo simulator.

Reviewer 2: ```suggestion
  This is a ROS 2 package for integrating the `ros2_control` controller architecture
  with the `Gazebo` simulator.
```

### How to prepare a Robot URDF for using Gazebo with ros2_control

> Two main actions needs to be done to prepare the URDF model to use Gazebo and
> ros2_control: configure ros2_control setting using the `ros2_control` URDF
> macro, and add a plugin to Gazebo to parse the `ros2_control` tags and
> load the appropriate hardware interfaces and controller manager.

Reviewer 2: ```suggestion
Two main actions needs to be done to prepare the URDF model to use `Gazebo` and
`ros2_control`: configure `ros2_control` setting using the `ros2_control` URDF
macro, and add a plugin to `Gazebo` to parse the `ros2_control` tags and
load the appropriate hardware interfaces and controller manager.
```

#### 1. URDF declaration for the [`ros2_control` URDF tag](https://control.ros.org/master/doc/getting_started/getting_started.html#hardware-description-in-urdf)

This is done by adding the xacro macro `ros2_control_panda_arm` (which is
integrated from main Panda arm URDF file) and using the
[`ign_ros2_control/IgnitionSystem`](https://github.com/ros-controls/gz_ros2_control/blob/master/README.md?plain=1#L93-L118):

* https://github.com/AndrejOrsula/panda_ign_moveit2/blob/master/panda_description/urdf/panda.ros2_control#L28-L30

#### 2. URDF declaration for the Gazebo plugin [IgnitionROS2ControlPlugin](https://github.com/ros-controls/gz_ros2_control/blob/master/README.md?plain=1#L153-L169)

> This is done by adding the xacro macro `ign_ros2_control` (which is integrated
> from main Panda URDF file) and adding the Gazebo plugin `IgnitionROS2ControlPlugin`
> that parses the `ros2_control` tags and loads the appropriate hardware interfaces and controller manager

Reviewer 2: ```suggestion
This is done by adding the xacro macro `gz_ros2_control` (which is integrated
from main Panda URDF file) and adding the Gazebo plugin `IgnitionROS2ControlPlugin`
that parses the `ros2_control` tags and loads the appropriate hardware interfaces and controller manager
```

`IgnitionROS2ControlPlugin` will be `GazeboRos2ControlPlugin`

* Panda main MoveIt 2 move_group invocation:
  * https://github.com/AndrejOrsula/panda_ign_moveit2/blob/5679c0a97d5e34199271ed7abf02cf303d0e8fb9/panda_moveit_config/launch/move_group.launch.py#L248-L264

* Controller manager
  * https://github.com/AndrejOrsula/panda_ign_moveit2/blob/5679c0a97d5e34199271ed7abf02cf303d0e8fb9/panda_moveit_config/launch/move_group.launch.py#L309-L324

### How the Follow Target example declares the target pose for the manipulation

This repository uses a `Gazebo PosePublisher` plugin to indicate the target pose for the manipulation:
* [follow_target.sdf](../worlds/follow_target.sdf#L99-L104)

Translate the pose from Gazebo to ROS using the `ros_ign_bridge`:
* [world_follow_target.launch.py](../launch/worlds/world_follow_target.launch.py#L61-L76)

The `/target_pose` topic is subscribed to by the ROS C++ node that communicates with MoveIt
* [ex_follow_target.cpp](../examples/cpp/ex_follow_target.cpp#L28-L57)

### Other important ROS 2 packages

#### Robot state publisher

> The Robot State Publisher is a node and a class to publish
> the state of a robot to tf2. At startup time, Robot State Publisher is
> supplied with a kinematic tree model (URDF) of the robot. It then subscribes
> to the `joint_states` topic (of type `sensor_msgs/msg/JointState`) to get
> individual joint states. More information in the
> [ROS 2 tutorial](https://docs.ros.org/en/galactic/Tutorials/Intermediate/URDF/Using-URDF-with-Robot-State-Publisher.html)

Reviewer 2: ```suggestion
The Robot State Publisher is a node and a class to publish
the state of a robot to `tf2`. At startup time, Robot State Publisher is
supplied with a kinematic tree model (URDF) of the robot. It then subscribes
to the `joint_states` topic (of type `sensor_msgs/msg/JointState`) to get
individual joint states. More information in the
[ROS 2 tutorial](https://docs.ros.org/en/galactic/Tutorials/Intermediate/URDF/Using-URDF-with-Robot-State-Publisher.html)
```

* https://github.com/ros/robot_state_publisher

#### TF2

> tf2 is the second generation of the transform library, which lets the user
> keep track of multiple coordinate frames over time. tf2 maintains the
> relationship between coordinate frames in a tree structure buffered in
> time, and lets the user transform points, vectors, etc between any
> two coordinate frames at any desired point in time.
> 
> * http://wiki.ros.org/tf2
> 
> 
> ## ROS 2 Topics and Nodes

Reviewer 2: ```suggestion
`tf2` is the second generation of the transform library, which lets the user
keep track of multiple coordinate frames over time. `tf2` maintains the
relationship between coordinate frames in a tree structure buffered in
time, and lets the user transform points, vectors, etc between any
two coordinate frames at any desired point in time.

* http://wiki.ros.org/tf2

## ROS 2 Topics and Nodes
```

# Example documentation

This document describes the different software components
used by the MoveIt 2, ROS 2 and Gazebo demo,
where to find them and how they are being called.

This example and documentation uses the name ***"Gazebo"***
to refer to the new generation of the [Gazebo simulator](https://gazebosim.org/)

> not be confused with [Gazebo classic](https://classic.gazebosim.org) also
> known as Gazebo11. Current "Gazebo" was first known as
> Ignition Gazebo so the document contains some references to "ign_" or
> "ignition" like in the name of the repository.

Reviewer: ```suggestion
not be confused with [Gazebo classic](https://classic.gazebosim.org), the latest
version of which is Gazebo 11. The current "Gazebo" was formerly known as
Ignition Gazebo, so the document contains some references to "ign_" or
"ignition", like in the name of the repository.
```

The Ignition Gazebo project was renamed to just

> Gazebo or Gz so the document contains some references to the new URLs
> and name like the ones for
> [`gz_ros2_control`](https://github.com/ros-controls/gz_ros2_control)
> that should not be confused with the `ros2_control` wrapper for the
> previous Gazebo classic named
> [`gazebo_ros2_control`](https://github.com/ros-controls/gazebo_ros2_control).

Reviewer: ```suggestion
Gazebo or Gz, so the document contains some references to the new URLs
and names, like the ones for
[`gz_ros2_control`](https://github.com/ros-controls/gz_ros2_control),
which should not be confused with
[`gazebo_ros2_control`](https://github.com/ros-controls/gazebo_ros2_control),
the `ros2_control` wrapper for the previous Gazebo classic.
```
